File: 09-HASHING/Go/design-hashset.go
package main

// https://leetcode.com/problems/design-hashset/

type Node struct {
	key  int
	next *Node
}

type MyHashSet struct {
	set [10007]*Node
}

func Constructor() MyHashSet {
	return MyHashSet{}
}

func (this *MyHashSet) Add(key int) {
	tail, curr := this.get(key)
	if curr != nil {
		return
	}
	node := &Node{key: key}
	if tail == nil {
		this.set[this.index(key)] = node
	} else {
		tail.next = node
	}
}

func (this *MyHashSet) Remove(key int) {
	prev, curr := this.get(key)
	if curr == nil {
		return
	}

	if prev == nil {
		this.set[this.index(key)] = curr.next
	} else {
		prev.next = curr.next
	}
}

func (this *MyHashSet) Contains(key int) bool {
	_, curr := this.get(key)
	return curr != nil
}

func (this *MyHashSet) get(key int) (*Node, *Node) {
	var prev *Node
	curr := this.set[this.index(key)]
	for curr != nil {
		if curr.key == key {
			return prev, curr
		}
		prev = curr
		curr = curr.next
	}
	return prev, nil
}

func (this *MyHashSet) index(key int) int {
	return key % 10007
}

File: 09-HASHING/Python/design-hashset.py
# https://leetcode.com/problems/design-hashset/

class Node:
    def __init__(self, key: int):
        self.key = key
        self.next = None


class MyHashSet:
    size = 10007

    def __init__(self):
        self.set = [None] * self.size

    def add(self, key: int) -> None:
        tail, curr = self._get(key)
        if curr is not None:
            return

        node = Node(key=key)
        if tail is None:
            self.set[self._index(key)] = node
        else:
            tail.next = node

    def remove(self, key: int) -> None:
        prev, curr = self._get(key)
        if curr is None:
            return

        if prev is None:
            self.set[self._index(key)] = curr.next
        else:
            prev.next = curr.next

    def contains(self, key: int) -> bool:
        return self._get(key)[1] is not None

    def _index(self, key: int) -> int:
        return key % self.size

    def _get(self, key: int):
        index = self._index(key)
        curr = self.set[index]
        prev = None
        while curr:
            if curr.key == key:
                return prev, curr
            prev = curr
            curr = curr.next
        return prev, None
